dsgrid/config/registration_models.py

```diff
@@ -0,0 +1,129 @@
+"""Contains data models to control registration of test projects and datasets."""
+
+from collections.abc import Iterable
+from pathlib import Path
+from typing import Any, Optional
+
+from pydantic import Field, model_validator
+from typing_extensions import Annotated
+
+from dsgrid.data_models import DSGBaseModel
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.utils.files import load_data
+
+
+class DatasetRegistrationModel(DSGBaseModel):
+    """Defines a dataset to be registered and optionally submitted to a project."""
+
+    dataset_id: Annotated[str, Field(description="Dataset ID")]
+    dataset_path: Annotated[
+        Path, Field(description="Directory containing load_data/load_data_lookup.parquet")
+    ]
+    config_file: Annotated[Path, Field(description="Path to dataset.json5")]
+    dimension_mapping_file: Annotated[
+        Optional[Path],
+        Field(
+            description="Path to file containing mappings of dataset-to-project dimensions",
+            default=None,
+        ),
+    ]
+    dimension_mapping_references_file: Annotated[
+        Optional[Path],
+        Field(
+            description="Path to file containing references to mappings of dataset-to-project dimensions",
+            default=None,
+        ),
+    ]
+    replace_dimension_names_with_ids: Annotated[
+        bool,
+        Field(
+            description="Replace the dimension entries with IDs of dimensions in the database "
+            "with matching names. Typically only useful for tests.",
+            default=False,
+        ),
+    ]
+    replace_dimension_mapping_names_with_ids: Annotated[
+        bool,
+        Field(
+            description="Replace the dimension mapping entries with IDs of dimension mappings "
+            "in the database with matching names. Typically only useful for tests.",
+            default=False,
+        ),
+    ]
+    register_dataset: Annotated[
+        bool, Field(description="If True, register the dataset", default=True)
+    ]
+    submit_to_project: Annotated[
+        bool, Field(description="If True, submit the dataset to the project", default=True)
+    ]
+    autogen_reverse_supplemental_mappings: Annotated[
+        set[DimensionType],
+        Field(
+            description="Dimensions on which to attempt create reverse mappings from supplemental dimensions.",
+            default=set(),
+        ),
+    ]
```

Comment:
When do we do this/why would a user want to do this?

Reply:
This is the help provided by the `submit-dataset` CLI command:
```
│    --autogen-reverse-supplemental-mappings  -a  [metric|geography|sector|subsector|time|weather_year|model_year|scenario]  For any dimension listed here, if the dataset's dimension is a project's supplemental dimension and no mapping is provided,       │
│                                                                                                                            create a reverse mapping from that supplemental dimension. 
```